index.d.ts

```diff
@@ -1,288 +1,472 @@
 declare module 'react-table' {
-  import { ReactNode, useState } from 'react'
+  import { ReactNode, useState, ComponentType, MouseEvent } from 'react'
+
+  export interface TableOptions<D extends object>
+    extends UseTableOptions<D> { //,
+    // UseSortByOptions<D>,
+    // UseFiltersOptions<D>,
+    // UseGroupByOptions<D>,
+    // UseExpandedOptions<D>,
+    // UsePaginationOptions<D>,
+    // UseRowSelectOptions<D>,
+    // UseRowStateOptions<D> {
+    // columns: [] // FIXME
+    // state: [] //FIXME
+  }
 
-  type StringKey<D> = Extract<keyof D, string>
-  type IdType<D> = StringKey<D> | string
+  export interface TableInstance<D extends object = {}> extends TableOptions<D>, UseTableInstanceProps<D> {}
 
-  type SortingRule<D> = {
-    id: IdType<D>
-    desc: boolean
+  /**
+   * The empty definition of TableState provides a definition for the state, that can then be extended in the users code.
+   * 
+   * @example
+   *  export interface TableState<D extends object = {}}>
+   *    extends UsePaginationState,
+   *      UseGroupByState,
+   *      UseSortbyState<D>,
+   *      UseFiltersState<D> {}
+   */      
+  export interface TableState<D extends object = {}> {} // eslint-disable-line @typescript-eslint/no-empty-interface
+
+  export type Hooks<D extends object = {}> =
+    UseTableHooks<D> //&
+    // UseSortByHooks<D> &
+    // UseGroupByHooks<D> &
+    // UseExpandedHooks<D> &
+    // UseRowSelectHooks<D>
+
+  export type Cell<D extends object = {}> = UseTableCellProps<D>
+
+  export type Column<D extends object = {}> = UseTableColumnOptions<D>
+
+  export type ColumnInstance<D extends object = {}> = Column<D> & UseTableColumnProps<D>
+
+  export type HeaderGroup<D extends object = {}> = UseTableHeaderGroupProps<D>
+
+  export type Row<D extends object = {}> = UseTableRowProps<D>
+
+
+  /* #region useTable */
+  export function useTable<D extends object = {}>(
+    options: TableOptions<D>,
+    ...plugins: PluginHook<D>[] 
+  ): TableInstance<D>
+  
+  export type UseTableOptions<D extends object> = Partial<{
+    // columns: Column<D>[] // FIXME
```

Comment:
I think that this is why the old version had that weird split between Column and EnhancedColumn.

Column is that small subset of what is available to the user when defining columns, but EnhancedColumn contains all of the additional information that the library adds to columns that are passed through the rest of the api.

I do think that you are correct that it shouldn't extend TableInstance, that was my mistake.

Reply:
Right, I broke it into ColumnOptions and ColumnProps. Still trying to figure out how all these pieces go together easily for the composition.

Reply:
yeah, there's a lot to unpack in your changes :)

Reply:
related to this, I think that an awful lot of the Column references in here should actually be ColumnInstance instead.  At least staring with the definition in UseTableInstanceProps - Still working though the others - it's pretty much the replacement for EnhancedColumn, and that was correctly used all over the place.